File: impl/python/weavepack_ast/apply.py
"""weavepack-ast — delta application (init_state, apply_chain).

AST state:
  {
    'nodes': dict[str(nid), {
        'kind': str,
        'parentNid': int|None,
        'childIndex': int,
        'props': dict[int|str, {'ctype': int, 'value': any}],
    }],
  }

Profile isolation: imports only from .types.
Mirrors sdk/src/profiles/ast/apply.js exactly, including the
null-value-store behavior of prop_set (value=None is stored, not deleted).
"""

import copy
from .types import OP, PATH_KIND


def _collect_descendants(nodes: dict, root_nid: int) -> list:
    result = []
    queue = [root_nid]
    while queue:
        nid = queue.pop()
        for k, node in nodes.items():
            if node["parentNid"] == nid and int(k) not in result:
                result.append(int(k))
                queue.append(int(k))
    return result


def _populate_nodes_from_block(nodes: dict, block: dict):
    btype = block.get("type")
    nids = block.get("nids") or []
    parent_nids = block.get("parentNids") or []
    child_indices = block.get("childIndices") or []
    columns = block.get("columns") or []

    for i, nid in enumerate(nids):
        key = str(nid)
        if key in nodes:
            raise ValueError(f"duplicate_element_id: nid {nid} already exists")
        kind = block.get("kind", "") if btype == "node" else (block.get("kinds") or [])[i]
        props = {}
        for col in columns:
            v = col["values"][i]
            if v is not None:
                props[col["colId"]] = {"ctype": col["ctype"], "value": v}
        nodes[key] = {
            "kind": kind,
            "parentNid": parent_nids[i],
            "childIndex": child_indices[i],
            "props": props,
        }


def init_state(doc: dict) -> dict:
    """Build initial AST state from a decoded tree document."""
    state = {"nodes": {}}
    for blk in (doc.get("blocks") or []):
        _populate_nodes_from_block(state["nodes"], blk)
    return state


def apply_op(state: dict, op: dict) -> dict:
    """Apply a single op to AST state. Returns a new state (deep-copied)."""
    state = copy.deepcopy(state)
    op_code = op["op"]
    nodes = state["nodes"]

    if op_code == OP.NODE_INSERT:
        _populate_nodes_from_block(nodes, op["block"])

    elif op_code == OP.NODE_DELETE:
        for nid in (op.get("nids") or []):
            nid_int = int(nid)
            descendants = _collect_descendants(nodes, nid_int)
            nodes.pop(str(nid_int), None)
            for dk in descendants:
                nodes.pop(str(dk), None)

    elif op_code == OP.NODE_MOVE:
        nid_str = str(int(op["nid"]))
        node = nodes.get(nid_str)
        if node is None:
            raise ValueError(f"element_not_found: node {op['nid']} not found")
        new_parent = op.get("newParentNid", 0)
        node["parentNid"] = None if (new_parent == 0 or new_parent is None) else int(new_parent)
        node["childIndex"] = int(op.get("newChildIndex", 0))

    elif op_code == OP.PROP_SET:
        path = op["path"]
        kind = path["kind"]
        if kind == PATH_KIND.NODE_COL:
            col_key = path["colId"]
            nid_str = str(int(path["nid"]))
        elif kind == PATH_KIND.NODE:
            col_key = None
            nid_str = str(int(path["nid"]))
        elif kind == PATH_KIND.NODE_PROP:
            col_key = path.get("prop")
            nid_str = str(int(path["nid"]))
        else:
            return state  # no-op for other path kinds

        node = nodes.get(nid_str)
        if node is None:
            return state  # no-op if node not found

        if col_key is not None:
            # Store the value (including None/null) — mirrors astSpecToOp+applyOp in JS
            # which stores null rather than deleting the prop entry.
            node["props"][col_key] = {"ctype": op["ctype"], "value": op.get("value")}

    elif op_code == OP.KIND_RENAME:
        old_kind = op.get("oldKind", "")
        new_kind = op.get("newKind", "")
        for node in nodes.values():
            if node["kind"] == old_kind:
                node["kind"] = new_kind

    elif op_code == OP.SUBTREE_REPLACE:
        root_nid = int(op["rootNid"])
        descendants = _collect_descendants(nodes, root_nid)
        for dk in descendants:
            nodes.pop(str(dk), None)
        _populate_nodes_from_block(nodes, op["block"])

    else:
        raise ValueError(f"unknown_delta_op: op code {op_code}")

    return state


def apply_chain(state: dict, ops: list) -> dict:
    """Apply a list of ops in order. Returns the final AST state."""
    for op in ops:
        state = apply_op(state, op)
    return state

File: impl/rust/weavepack-ast/src/apply.rs
// weavepack-ast — delta application.
//
// Mirrors the logic in sdk/src/profiles/ast/apply.js.
// Profile isolation: only imports from crate::types.

use std::collections::BTreeMap;

use crate::types::{AstDoc, Block, CellValue, Op, Path};

// ── Col-id key (numeric or named) ──────────────────────────────────────────

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum ColId {
    Num(u32),
    Named(String),
}

// ── Per-node property entry ─────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct PropEntry {
    pub ctype: u8,
    pub value: Option<CellValue>,
}

// ── Node state ──────────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct NodeRecord {
    pub kind:        String,
    pub parent_nid:  Option<u64>,
    pub child_index: u32,
    pub props:       BTreeMap<ColId, PropEntry>,
}

// ── AST live state ──────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct AstState {
    pub schema_hash: [u8; 32],
    pub nodes:       BTreeMap<u64, NodeRecord>,
}

// ── Helpers ─────────────────────────────────────────────────────────────────

fn collect_descendants(nodes: &BTreeMap<u64, NodeRecord>, root: u64) -> Vec<u64> {
    let mut result = Vec::new();
    let mut queue  = vec![root];
    while let Some(nid) = queue.pop() {
        for (&child_nid, child) in nodes {
            if child.parent_nid == Some(nid) && !result.contains(&child_nid) {
                result.push(child_nid);
                queue.push(child_nid);
            }
        }
    }
    result
}

fn populate_nodes_from_block(nodes: &mut BTreeMap<u64, NodeRecord>, block: &Block) -> Result<(), String> {
    let (nids, parent_nids, child_indices, columns) = (
        block.nids(), block.parent_nids(), block.child_indices(), block.columns()
    );
    for i in 0..nids.len() {
        let nid = nids[i];
        if nodes.contains_key(&nid) {
            return Err(format!("duplicate_element_id: nid {nid} already exists"));
        }
        let kind = match block {
            Block::Node(b) => b.kind.clone(),
            Block::Mixed(b) => b.kinds[i].clone(),
        };
        let mut props = BTreeMap::new();
        for col in columns {
            if let Some(val) = &col.values[i] {
                props.insert(ColId::Num(col.col_id), PropEntry { ctype: col.ctype, value: Some(val.clone()) });
            }
        }
        nodes.insert(nid, NodeRecord {
            kind,
            parent_nid:  parent_nids[i],
            child_index: child_indices[i],
            props,
        });
    }
    Ok(())
}

// ── init_state ──────────────────────────────────────────────────────────────

pub fn init_state(doc: &AstDoc) -> Result<AstState, String> {
    let mut state = AstState {
        schema_hash: doc.schema_hash,
        nodes: BTreeMap::new(),
    };
    for block in &doc.blocks {
        populate_nodes_from_block(&mut state.nodes, block)?;
    }
    Ok(state)
}

// ── Op application ──────────────────────────────────────────────────────────

fn apply_op(state: &mut AstState, op: &Op) -> Result<(), String> {
    match op {
        Op::NodeInsert { block } => {
            populate_nodes_from_block(&mut state.nodes, block)?;
        }

        Op::NodeDelete { nids } => {
            for &nid in nids {
                let descendants = collect_descendants(&state.nodes, nid);
                state.nodes.remove(&nid);
                for dk in descendants {
                    state.nodes.remove(&dk);
                }
            }
        }

        Op::NodeMove { nid, new_parent_nid, new_child_index } => {
            let node = state.nodes.get_mut(nid)
                .ok_or_else(|| format!("element_not_found: node {nid} not found"))?;
            node.parent_nid  = if *new_parent_nid == 0 { None } else { Some(*new_parent_nid) };
            node.child_index = *new_child_index;
        }

        Op::PropSet { path, ctype, nullable: _, is_null: _, value } => {
            let (node_nid, col_id_key) = match path {
                Path::NodeCol { nid, col_id } => (*nid, Some(ColId::Num(*col_id))),
                Path::Node { nid }             => (*nid, None),
                Path::NodeProp { nid, prop }   => (*nid, Some(ColId::Named(prop.clone()))),
                _                              => return Ok(()),
            };
            let node = match state.nodes.get_mut(&node_nid) {
                Some(n) => n,
                None    => return Ok(()), // no-op if node not found
            };
            if let Some(key) = col_id_key {
                // Store the value (including None/null) — mirrors astSpecToOp+applyOp in JS
                // which stores null rather than deleting the prop entry.
                node.props.insert(key, PropEntry { ctype: *ctype, value: value.clone() });
            }
        }

        Op::KindRename { old_kind, new_kind } => {
            for node in state.nodes.values_mut() {
                if node.kind == *old_kind {
                    node.kind = new_kind.clone();
                }
            }
        }

        Op::SubtreeReplace { root_nid, block } => {
            // Delete all descendants of root_nid (keep root itself)
            let descendants = collect_descendants(&state.nodes, *root_nid);
            for dk in descendants {
                state.nodes.remove(&dk);
            }
            // Insert replacement subtree
            populate_nodes_from_block(&mut state.nodes, block)?;
        }
    }
    Ok(())
}

// ── apply_chain ─────────────────────────────────────────────────────────────

pub fn apply_chain(mut state: AstState, ops: &[Op]) -> Result<AstState, String> {
    for op in ops {
        apply_op(&mut state, op)?;
    }
    Ok(state)
}
